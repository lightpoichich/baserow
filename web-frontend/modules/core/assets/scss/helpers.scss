.clearfix::after {
  clear: both;
  content: '';
  display: table;
}

.hidden {
  display: none;
}

.visibility-hidden {
  visibility: hidden;
  transform: translateX(-10000px) translateY(-10000px);
}

.align-right {
  text-align: right;
}

.color-primary {
  color: $color-primary-500 !important;
}

.margin-top-0 {
  margin-top: 0 !important;
}

.margin-top-1 {
  margin-top: 8px !important;
}

.margin-top-2 {
  margin-top: 16px !important;
}

.margin-top-3 {
  margin-top: 30px !important;
}

.margin-top-4 {
  margin-top: 40px !important;
}

.margin-bottom-0 {
  margin-bottom: 0 !important;
}

.margin-bottom-1 {
  margin-bottom: 8px !important;
}

.margin-bottom-2 {
  margin-bottom: 16px !important;
}

.margin-bottom-3 {
  margin-bottom: 30px !important;
}

.margin-bottm-4 {
  margin-bottom: 40px !important;
}

.margin-right-1 {
  margin-right: 8px;
}

.resizing-horizontal {
  cursor: col-resize;
}

// This selector can only be used to temporarily show an element for calculating
// positions.
.forced-block {
  display: block !important;
}

// Sometimes we forcefully want to disable the the user-select property to initial
// because the element is for example being edited via a contenteditable attribute
// which does not work in combination with the user-select: none; property in Safari.
.forced-user-select-initial {
  user-select: initial !important;
}

.prevent-scroll {
  overflow: hidden !important;
}

@keyframes spin {
  0% {
    transform: rotate(0);
  }

  50% {
    transform: rotate(360deg);
  }

  100% {
    transform: rotate(720deg);
  }
}
