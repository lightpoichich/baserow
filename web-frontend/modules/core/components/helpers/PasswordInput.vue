<!--
This component can be used in a form which requires password validation.
The password requirements for baserow are encapsulated in this component.

In order to use this component in another component/form it needs a passwordValue
prop. That means the state for the passwordValue always lives in a parent component.
The parent component needs to provide a function in order to update the state.
When the input is changed this component emits an 'inputChange' event with the event.

In order to access the validation of this component in the parent component a 'ref'
needs to be added to the component. With that the vuelidate instance can be accessed:

e.g. this.$refs.REFNAME.$v

Example Usage:

<PasswordInput
   ref="password"
   label="New Password"
   name="password"
   :password-value="PASSWORDSTATE"
   @inputChange="handleChange"
/>

-->
<template>
  <div>
    <label class="control__label">{{ label }}</label>
    <div class="control__elements">
      <input
        :class="{ 'input--error': $v.passwordValue.$error }"
        :name="name"
        :value="passwordValue"
        type="password"
        class="input input--large"
        @blur="$v.passwordValue.$touch()"
        @input="(event) => this.$emit('inputChange', event)"
      />
      <div
        v-if="$v.passwordValue.$error && !$v.passwordValue.required"
        class="error"
      >
        Input is required.
      </div>
      <div
        v-if="$v.passwordValue.$error && !$v.passwordValue.maxLength"
        class="error"
      >
        A maximum of
        {{ $v.passwordValue.$params.maxLength.max }} characters is allowed here.
      </div>
      <div
        v-if="$v.passwordValue.$error && !$v.passwordValue.minLength"
        class="error"
      >
        A minimum of
        {{ $v.passwordValue.$params.minLength.min }} characters is required
        here.
      </div>
    </div>
  </div>
</template>

<script>
import { maxLength, minLength, required } from 'vuelidate/lib/validators'

export default {
  name: 'PasswordInput',
  props: {
    label: {
      type: String,
      required: true,
    },
    passwordValue: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      required: true,
    },
  },
  methods: {
    isFormValid() {
      this.$v.passwordValue.$touch()
      return !this.$v.$invalid
    },
  },
  validations: {
    passwordValue: {
      required,
      maxLength: maxLength(256),
      minLength: minLength(8),
    },
  },
}
</script>
