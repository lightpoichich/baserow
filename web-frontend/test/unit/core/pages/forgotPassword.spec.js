import { TestApp } from '@baserow/test/helpers/testApp'
import forgotPasswordPage from '@baserow/modules/core/pages/forgotPassword'

class ForgotPasswordPageHelpers {
  constructor(forgotPasswordPageComponent) {
    this.c = forgotPasswordPageComponent
  }

  findForm() {
    return this.c.find('form')
  }

  findEmailInput() {
    return this.c.find('form input')
  }

  errorEmailRequired() {
    return this.c.vm.v$.email.required.$invalid
  }

  errorEmailInvalid() {
    return this.c.vm.v$.email.email.$invalid
  }

  async setEmailInputValue(value) {
    await this.findEmailInput().setValue(value)
    await this.findEmailInput().trigger('input')
    await this.forceUpdate()
  }

  async triggerEmailInputBlur() {
    await this.findEmailInput().trigger('blur')
    await this.c.vm.v$.$touch()
    await this.forceUpdate()
  }

  async triggerSubmit() {
    await this.findForm().trigger('submit.prevent')
    await this.forceUpdate()
  }

  isInvalidEmailErrorMessageShown() {
    return this.c.html().includes('error.invalidEmail')
  }

  async forceUpdate() {
    await this.c.vm.$nextTick()
    await this.c.vm.$forceUpdate()
  }
}

describe('forgotPassword page', () => {
  let testApp = null

  beforeEach(() => {
    testApp = new TestApp()
  })

  afterEach(() => testApp.afterEach())

  async function whenTheForgotPasswordFormIsShown() {
    testApp.store.commit('settings/SET_SETTINGS', {
      allow_reset_password: true,
    })
    const page = await testApp.mount(forgotPasswordPage, { sync: false })
    const helper = new ForgotPasswordPageHelpers(page)
    return { page, helper }
  }

  test('Default form match snapshot', async () => {
    const { page } = await whenTheForgotPasswordFormIsShown()
    expect(page.element).toMatchSnapshot()
  })

  test('A form is rendered', async () => {
    const { helper } = await whenTheForgotPasswordFormIsShown()
    // Ensure the form exists
    expect(helper.findForm().exists()).toBe(true)
  })

  test('An error is shown if the email is empty', async () => {
    const { helper } = await whenTheForgotPasswordFormIsShown()

    expect(helper.isInvalidEmailErrorMessageShown()).toBe(false)

    await helper.triggerSubmit()

    expect(helper.errorEmailRequired()).toBe(true)
    expect(helper.errorEmailInvalid()).toBe(false)
    expect(helper.isInvalidEmailErrorMessageShown()).toBe(true)
  })

  test('An error is shown if the email is not valid', async () => {
    const { helper } = await whenTheForgotPasswordFormIsShown()

    expect(helper.isInvalidEmailErrorMessageShown()).toBe(false)

    await helper.setEmailInputValue('invalid-email')
    await helper.triggerEmailInputBlur()

    expect(helper.errorEmailRequired()).toBe(false)
    expect(helper.errorEmailInvalid()).toBe(true)
    expect(helper.isInvalidEmailErrorMessageShown()).toBe(true)
  })

  test('No error is shown with a valid email', async () => {
    const { helper } = await whenTheForgotPasswordFormIsShown()

    await helper.setEmailInputValue('valid@email.com')
    await helper.triggerEmailInputBlur()

    expect(helper.errorEmailRequired()).toBe(false)
    expect(helper.errorEmailInvalid()).toBe(false)
    expect(helper.isInvalidEmailErrorMessageShown()).toBe(false)
  })
})
