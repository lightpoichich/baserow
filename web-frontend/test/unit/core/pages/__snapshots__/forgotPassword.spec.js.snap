// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forgotPassword page Default form 1`] = `
<div>
  <div>
    <div
      class="auth__logo"
    >
      <a
        to="[object Object]"
      >
        <div
          class="logo"
        >
          <img
            class=""
            src="@baserow/modules/core/static/img/logo.svg"
          />
        </div>
      </a>
    </div>
     
    <div
      class="auth__head auth__head--more-margin"
    >
      <h1
        class="auth__head-title"
      >
        forgotPassword.title
      </h1>
       
      <div
        class="lang-picker"
      >
        <a
          class="lang-picker__link"
        >
          
    
    
          <i
            class="iconoir-nav-arrow-down lang-picker__icon"
          />
        </a>
         
        <div
          class="dropdown lang-picker__dropdown dropdown--floating"
          tabindex="0"
        >
          <!---->
           
          <div
            class="dropdown__items hidden"
          >
            <!---->
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item select__item--no-options"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                     
                    <!---->
                     
                    <!---->
                     
                    <span
                      class="select__item-name-text"
                      title="English"
                    >
                      English
                    </span>
                  </div>
                   
                  <!---->
                </a>
                 
                <i
                  class="select__item-active-icon iconoir-check"
                />
              </li>
            </ul>
             
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="alert alert--error"
    >
      <i
        class="alert__icon iconoir-warning-circle"
      />
       
      <div
        class="alert__content"
      >
        <div
          class="alert__title"
        >
          forgotPassword.disabled
        </div>
         
        <div
          class="alert__message"
        >
           
          <p>
            forgotPassword.disabledMessage
          </p>
        </div>
         
        <!---->
      </div>
       
      <!---->
    </div>
     
    <a
      class="button button--full-width"
      to="[object Object]"
    >
      <i
        class="iconoir-arrow-left"
      />
      
        action.backToLogin
      
    </a>
  </div>
   
  <!---->
</div>
`;

exports[`forgotPassword page Default form match snapshot 1`] = `
<div>
  <div>
    <div
      class="auth__logo"
    >
      <a
        to="[object Object]"
      >
        <div
          class="logo"
        >
          <img
            class=""
            src="@baserow/modules/core/static/img/logo.svg"
          />
        </div>
      </a>
    </div>
     
    <div
      class="auth__head auth__head--more-margin"
    >
      <h1
        class="auth__head-title"
      >
        forgotPassword.title
      </h1>
       
      <div
        class="lang-picker"
      >
        <a
          class="lang-picker__link"
        >
          
    
    
          <i
            class="iconoir-nav-arrow-down lang-picker__icon"
          />
        </a>
         
        <div
          class="dropdown lang-picker__dropdown dropdown--floating"
          tabindex="0"
        >
          <!---->
           
          <div
            class="dropdown__items hidden"
          >
            <!---->
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item select__item--no-options"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                     
                    <!---->
                     
                    <!---->
                     
                    <span
                      class="select__item-name-text"
                      title="English"
                    >
                      English
                    </span>
                  </div>
                   
                  <!---->
                </a>
                 
                <i
                  class="select__item-active-icon iconoir-check"
                />
              </li>
            </ul>
             
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
    </div>
     
    <div>
      <p>
        
        forgotPassword.message
      
      </p>
       
      <!---->
       
      <form>
        <div
          class="auth__control"
        >
          <label
            class="auth__control-label"
          >
            field.emailAddress
          </label>
           
          <div
            class="control__elements"
          >
            <input
              class="input"
              type="text"
            />
             
            <div
              class="auth__control-error"
            >
              <!---->
            </div>
          </div>
        </div>
         
        <div
          class="auth__action"
        >
          <button
            class="button button--large button--full-width"
          >
            <!---->
             
            <span
              class="button__label"
            >
              
            forgotPassword.submit
          
            </span>
             
            <!---->
          </button>
        </div>
         
        <div>
          <ul
            class="auth__action-links"
          >
            <li>
              
              forgotPassword.loginText
              
              <a
                to="[object Object]"
              >
                
                action.login
              
              </a>
            </li>
          </ul>
        </div>
      </form>
    </div>
  </div>
   
  <!---->
</div>
`;
